lib/seahorn/CexExeGenerator.cc

```diff
@@ -0,0 +1,568 @@
+
+  SmallVector<Constant *, 20> LLVMValueSegment;
+
+  // total block size in bytes;
+  expr::mpz_class sizeMpz;
+  size_t blockWidth = 0;
+  if (expr::numeric::getNum(size, sizeMpz)) {
+    blockWidth = sizeMpz.get_ui();
+  } else {
+    LOG("cex",
+        errs() << "memhavoc: cannot get concrete size (" << *size << ")\n");
+    ArrayType *placeholderT = ArrayType::get(Type::getInt8PtrTy(ctx), 0);
+    return ConstantArray::get(placeholderT, LLVMValueSegment);
+  }
+
+  // starting address
+  expr::mpz_class startAddrMpz;
+  size_t startAddrInt = 0;
+  if (expr::numeric::getNum(startAddr, startAddrMpz)) {
+    startAddrInt = startAddrMpz.get_ui();
+  } else {
+    LOG("cex", errs() << "memhavoc: cannot get concrete starting address: "
+                      << *startAddr << "\n");
+    ArrayType *placeholderT = ArrayType::get(Type::getInt8PtrTy(ctx), 0);
+    return ConstantArray::get(placeholderT, LLVMValueSegment);
+  }
+
+  const MemMap m_map(segment);
+  if (!m_map.isValid()) {
+    LOG("cex", errs() << "memhavoc: cannot extract content from: " << *segment
+                      << "\n");
+    ArrayType *placeholderT = ArrayType::get(Type::getInt8PtrTy(ctx), 0);
+    return ConstantArray::get(placeholderT, LLVMValueSegment);
+  }
+  size_t elmWidth = m_map.getContentWidth();
+  size_t blocks = std::ceil((float)blockWidth / (float)elmWidth);
+  auto *segmentElmTy = IntegerType::get(ctx, elmWidth * 8);
+  ArrayType *segmentAT = ArrayType::get(segmentElmTy, blocks);
+
+  Expr defaultE = m_map.getDefault();
+  // get default value or use 0 as fallback
+  Constant *defaultConst;
+  if (defaultE) {
+    defaultConst = exprToLlvm(segmentElmTy, defaultE, ctx, dl);
+  } else {
+    LOG("cex", errs() << "havocing mem with default 0 \n");
+    defaultConst = Constant::getIntegerValue(
+        segmentElmTy, APInt(dl.getTypeStoreSizeInBits(segmentElmTy), 0));
+  }
+
+  // fill with default values
+  for (size_t i = 0; i < blocks; i++) {
+    LLVMValueSegment.push_back(defaultConst);
+  }
+
+  // fill special value indicated by ID
+  for (auto begin = m_map.cbegin(), end = m_map.cend(); begin != end; begin++) {
+    Expr segmentID = begin->getIdxExpr();
+    expr::mpz_class idE = 0;
+    if (expr::numeric::getNum(segmentID, idE)) {
+      size_t curAddrInt = idE.get_ui();
+      size_t index = (curAddrInt - startAddrInt) / elmWidth;
+      if (index >= blocks) {
+        LOG("cex", errs() << "memhavoc: out of bound index: [" << index
+                          << "] with only " << blocks << " blocks \n");
+        continue;
+      }
+      Expr segmentE = begin->getValueExpr();
+      auto *segmentConst = exprToLlvm(segmentElmTy, segmentE, ctx, dl);
+      LLVMValueSegment[index] = segmentConst;
+    } else
+      continue;
+  }
+  return ConstantArray::get(segmentAT, LLVMValueSegment);
+}
+
+bool extractArrayContents(Expr e, DsaKeyVal &out, Expr &defaultValue) {
+  const MemMap a_map(e);
```

Comment:
note that this is rewritten to leverage the more robust `ExprMemMap`; reading older version in `CexHarness` I notices the process is nearly identical to `ExprMemMap`'s array visitor

Reply:
that's great